api-server/src/server/boot/user.js

```diff
@@ -93,8 +112,89 @@ function deleteUserTokenResponse(req, res) {
   return res.send({ userToken: null });
 }
 
+function createPostMsUsername(app) {
+  const { MsUsername } = app.models;
+
+  return async function postMsUsername(req, res) {
+    // example msTranscriptUrl: https://learn.microsoft.com/en-us/users/mot01/transcript/8u6awert43q1plo
+    // the last part is the transcript ID
+    // the username is irrelevant, and retrieved from the MS API response
+
+    const { msTranscriptUrl } = req.body;
+
+    if (!msTranscriptUrl) {
+      return res
+        .status(400)
+        .send('Please include a Microsoft transcript URL in request');
+    }
+
+    const msTranscriptId = msTranscriptUrl.split('/').pop();
+    const msTranscriptApiUrl = `https://learn.microsoft.com/api/profiles/transcript/share/${msTranscriptId}`;
+
+    try {
+      const msApiRes = await fetch(msTranscriptApiUrl);
+
+      if (!msApiRes.ok) {
+        res.status(500);
+        throw new Error(
+          'An error occurred trying to get your Microsoft transcript'
+        );
+      }
+
+      const { userName } = await msApiRes.json();
+
+      if (!userName) {
+        res.status(500);
+        throw new Error(
+          'An error occured trying to link your Microsoft account'
+        );
+      }
```

Comment:
Is this definitely a `5XX`? Could the username be _missing_ or `undefined` for another reason?

Reply:
I suppose I can't be 100% sure, but it seems like it shouldn't ever be missing. Transcripts can only be created by users, and users have to have a username (I tried deleting mine, didn't work) - when you check the transcript, you get the username every time as far as I can tell. I don't know exactly what the API is doing, or what's stored in their database - so it's not impossible. This seems appropriate.